### PR TITLE
Add firestore database interface

### PR DESCRIPTION
**Pull request by bbjubjub2494**
Pre-requisite: #4 

Supersedes: #11 

TODO
- [x] displayOn

### CODE REVIEW

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/Firestore/Model/DataSetInterface.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.github.HumanLearning2021.HumanLearningApp.Model.Category
+import com.github.HumanLearning2021.HumanLearningApp.Model.DataSetInterface
+import com.google.firebase.firestore.DocumentId
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.firestore
+import com.google.firebase.ktx.Firebase
+import kotlinx.coroutines.tasks.await
+
+sealed class FirestoreDocument(val self: DocumentReference) {
```
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/Firestore/Model/DataSetInterface.kt]: - [x] This hierarchy ought to be in its own file.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/Firestore/Model/DataSetInterface.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        TODO("Not yet implemented")
+    }
+}
+
+class DataSetInterface(
+    private val db: FirebaseFirestore = Firebase.firestore,
+    private val prefix: String = "/databases/demo",
+) : DataSetInterface {
+
+    class CategorySchema {
```
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/Firestore/Model/DataSetInterface.kt]: - [x] These need not be public

**Comment by bbjubjub2494** [comment]
- [x] I'll refactor both test views into one

**Comment by bbjubjub2494** [comment]
appears to be due to Glide...

**Comment by bbjubjub2494** [comment]
ok at least I have coverage data

**Review by wattsecs — APPROVED** [review]
Looks good to me.

**Review by martin-lenweiter — APPROVED** [review]
All good

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDisplayPictureActivityTest.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import com.github.HumanLearning2021.HumanLearningApp.view.DisplayImageActivity
+import com.schibsted.spain.barista.assertion.BaristaImageViewAssertions.assertHasAnyDrawable
+import kotlinx.coroutines.runBlocking
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class FirestoreDisplayImageActivityTest {
```
[the remark this answers is not in the document]
Comment (bbjubjub2494) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDisplayPictureActivityTest.kt]: The activity is now called Image. however it's not a firestore specific activity so I'm going to rethink this name from scratch.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import android.app.Activity
+import android.widget.ImageView
+import com.bumptech.glide.Glide
+import com.github.HumanLearning2021.HumanLearningApp.model.CategorizedPicture
+import com.github.HumanLearning2021.HumanLearningApp.model.Category
+import com.google.firebase.ktx.Firebase
+import com.google.firebase.storage.ktx.storage
+
+data class FirestoreCategorizedPicture(
+    val path: String,
```
[the remark this answers is not in the document]
Comment (bbjubjub2494) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt:12]: No this is the path of the document within the firestore database, so it's a very specific format, for instance `/databases/demo/pictures/LbaIwsl1kizvTod4q1TG`. Do you think this requires documentation?

[the remark this answers is not in the document]
Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreCategorizedPicture.kt:12]: Okay I see. Documentation of this and some other implementation details (I would have a few questions) could be good if someone else is to contribute to the Firestore part of the code, but the behavior of the functions is clear enough not to require documentation. So I wouldn't say it's a priority if we want to merge tonight.

File: app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDisplayPictureActivityTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (martin-lenweiter) [reply app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDisplayPictureActivityTest.kt]: Sounds good

**Comment by bbjubjub2494** [comment]
Hey codeclimate I didn't write these

**Pull request merged**